keps/sig-network/0031-20181017-kube-proxy-services-optional.md

```diff
@@ -0,0 +1,124 @@
+
+When this label is set, kube-proxy will behave as if that service does not exist. None of the functionality that kube-proxy provides will be available for that service.
+
+kube-proxy will properly implement this label both at object creation and on dynamic addition/removal/updates of this label, either providing functionality or not for the service based on the latest version on the object.
+
+It is expected that this feature will mainly be used on large clusters with lots (>1000) of services. Any use of this feature in a smaller cluster will have negligible impact.
+
+The envisioned cluster that will make use of this feature looks something like the following:
+* Most/all traffic from outside the cluster is handled by gateways, such that each service in the cluster does not need a nodePort
+* These small number of entry points into the cluster are a part of the service mesh
+* There are many micro-services in the cluster, all a part of the service mesh, that are only accessed from inside the service mesh
+
+Higher level frameworks built on top of service meshes, such as [Knative](https://github.com/knative/docs), will be able to enable this feature by default due to having a more controlled application/service model and being reliant on the service mesh.
+
+#### Design
+
+Currently, when ProxyServer starts up it creates informers for all Service (ServiceConfig) and Endpoints (EndpointsConfig) objects using a single shared informer factory.
+
+The new design will simply add a LabelSelector filter to the shared informer factory, such that objects with the above label are filtered out by the API server:
+```diff
+-       informerFactory := informers.NewSharedInformerFactory(s.Client, s.ConfigSyncPeriod)
++       informerFactory := informers.NewSharedInformerFactoryWithOptions(s.Client, s.ConfigSyncPeriod,
++               informers.WithTweakListOptions(func(options *v1meta.ListOptions) {
++                       options.LabelSelector = "!service.kubernetes.io/alternative-service-proxy"
++               }))
+```
+
+This code will also handle the dynamic label update case. When the label selector is matched (service is enabled) an 'add' event will be generated by the informer. When the label selector is not matched (service is disabled) a 'delete' event will be generated by the informer.
+
+#### Testing
+
+The following cases should be tested. In each case, make sure that services are added/removed from iptables (or other) as expected:
+* Adding/removing services/endpoints with and without the above label
+* Adding/removing the above label from existing services/endpoints
+* Having a label value other than 'true', which should behave as if the label is not set
+
+### Risks and Mitigations
+
+We will keep the existing behaviour enabled by default, and only disable it when the cluster operator specifically asks to do so.
```

Comment:
I would like to add an additional clarifier:

Alternate kube-proxy implementations may ignore this value and still remain conformant with kubernetes services.

(In the spirit of this being an optional flag)

Reply:
Added this to the KEP.